### PR TITLE
Avoid full scan of match when there is an Relational In predicate

### PR DESCRIPTION
**Pull request by czpmango**
## What type of PR is this?
- [ ] bug
- [ ] feature
- [x] enhancement

#### Issue(s) number: 
https://github.com/vesoft-inc/nebula/issues/4708

## Checklist:
Tests:
- [ ] Unit test(positive and negative cases)
- [ ] Function test
- [x] Performance test
- [ ] N/A

Affects:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [x] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

## Release notes:

Please confirm whether to be reflected in release notes and how to describe:
> ex. Fixed the bug .....

### CODE REVIEW

File: src/graph/util/ExpressionUtils.cpp [hunk 1]
```diff
@@ -177,6 +180,30 @@ Expression *ExpressionUtils::rewriteParameter(const Expression *expr, QueryConte
   return graph::RewriteVisitor::transform(expr, matcher, rewriter);
 }
 
+Expression *ExpressionUtils::rewriteInnerInExpr(const Expression *expr) {
```
Comment (Shylock-Hg) [inline src/graph/util/ExpressionUtils.cpp:183]: Why do this?

Comment (czpmango) [reply src/graph/util/ExpressionUtils.cpp:183]: Equivalently rewrite IN expression with only one operand into EQ expression to reuse index selection logic.

File: src/common/expression/ContainerExpression.h [hunk 2]
```diff
@@ -63,7 +66,16 @@ class MapItemList final {
   std::vector<Pair> items_;
 };
 
-class ListExpression final : public Expression {
+class ContainerExpression : public Expression {
```
Comment (czpmango) [inline src/common/expression/ContainerExpression.h:69]: Just minor refactoring to provide some generic polymorphic methods for ContainerExpression.

File: tests/tck/features/match/IndexSelecting.feature [hunk 3]
```diff
@@ -57,6 +57,20 @@ Feature: Index selecting for match statement
       | 2  | AppendVertices | 5            |                                                     |
       | 5  | IndexScan      | 0            | {"indexCtx": {"columnHints":{"scanType":"PREFIX"}}} |
       | 0  | Start          |              |                                                     |
+    When profiling query:
+      """
+      MATCH (v:player) WHERE v.player.name in ["Yao Ming"] RETURN v.player.name AS name
```
Comment (Shylock-Hg) [inline tests/tck/features/match/IndexSelecting.feature:62]: Why don't support list with more items.

**Review by Aiee — APPROVED** [review]

File: src/common/expression/ContainerExpression.h [hunk 4]
```diff
@@ -63,7 +66,16 @@ class MapItemList final {
...
+ public:
+  virtual const std::vector<Expression *> getKeys() const = 0;
```
Comment (Aiee) [inline src/common/expression/ContainerExpression.h:71]: `getItems()` or `getElements` seems to be a better name

**Review by jievince — APPROVED** [review]

File: tests/tck/features/match/IndexSelecting.feature (hunk 3 above)
Comment (czpmango) [reply tests/tck/features/match/IndexSelecting.feature:62]: Fixed.

Comment (Shylock-Hg) [reply tests/tck/features/match/IndexSelecting.feature:62]: Add related tests.

Comment (czpmango) [reply tests/tck/features/match/IndexSelecting.feature:62]: Already added.

File: src/graph/util/ExpressionUtils.cpp [hunk 5]
```diff
@@ -177,6 +180,39 @@ Expression *ExpressionUtils::rewriteParameter(const Expression *expr, QueryConte
...
+  auto matcher = [](const Expression *e) -> bool {
+    if (e->kind() != Expression::Kind::kRelIn) {
+      return false;
+    }
+    auto rhs = static_cast<const RelationalExpression *>(e)->right();
+    return rhs->kind() == Expression::Kind::kList || rhs->kind() == Expression::Kind::kSet;
```
Comment (Shylock-Hg) [inline src/graph/util/ExpressionUtils.cpp:189]: Maybe should limit size of List to avoid stack overflow.

Comment (czpmango) [reply src/graph/util/ExpressionUtils.cpp:189]: Where is the stack overflow?

**Review by Aiee — APPROVED** [review]

**Review by jievince — APPROVED** [review]

File: src/graph/util/ExpressionUtils.cpp (hunk 5 above)
Comment (czpmango) [reply src/graph/util/ExpressionUtils.cpp]: @Shylock-Hg Construct flatten LogicalExpression instead.

**Review by Shylock-Hg — APPROVED** [review]

**Review by Aiee — APPROVED** [review]

**Comment by jinyingsunny** [comment]
checked on 3.3-rc

**Pull request merged**